core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/VisitedNodes.kt

```diff
@@ -116,26 +178,58 @@ data class VisitedNodes(val minDelay: Double) {
     fun markAsVisited(
         parameters: Parameters,
     ) {
-        val visitedTimes =
-            visitedRangesPerLocation.getOrPut(parameters.fingerprint!!) { TreeRangeSet.create() }
-
-        // We still add some padding to the end of the range, to avoid evaluating trains that are
-        // close to one another separately
-        val endVisitedTime = parameters.startTime + parameters.duration + minDelay
-        val newRange = Range.closed(parameters.startTime, endVisitedTime)
-        visitedTimes.add(newRange)
-
-        if (parameters.maxMarginDuration > 0.0) {
-            val reachableRanges =
-                reachableRangesPerLocation.getOrPut(parameters.fingerprint!!) {
-                    val map = TreeRangeMap.create<Double, LinearFunction>()
-                    map.put(Range.all(), LinearFunction(POSITIVE_INFINITY))
-                    map
-                }
-            val reachableRange =
-                Range.closed(endVisitedTime, endVisitedTime + parameters.maxMarginDuration)
-            val value = LinearFunction(parameters.baseNodeCost - endVisitedTime)
-            reachableRanges.merge(reachableRange, value) { f, g -> min(f, g) }
+        val visitedRanges =
+            visitedRangesPerLocation.getOrPut(parameters.fingerprint!!) { TreeRangeMap.create() }
+
+        val timeData = parameters.timeData
+        val startTime = timeData.earliestReachableTime
+        val maxDepartureTimeChange =
+            min(
+                timeData.maxDepartureDelayingWithoutConflict,
+                timeData.stopTimeData.minOfOrNull { it.maxDepartureDelayBeforeStop }
+                    ?: Double.POSITIVE_INFINITY
+            )
+
+        // We still add some padding to the end of each range and value, to avoid evaluating trains
+        // that are close to one another separately (`minDelay`)
+        val endRangeDepartureTimeChange = startTime + maxDepartureTimeChange + minDelay
+        val endRangeExtraStopTime =
+            startTime + timeData.maxDepartureDelayingWithoutConflict + minDelay
+        val endRangeExtraTravelTime = endRangeExtraStopTime + parameters.maxMarginDuration
+
+        fun putRange(start: Double, end: Double, value: ConditionallyVisitedRange) {
+            if (start < end) {
+                val range = Range.closedOpen(start, end)
+                visitedRanges.merge(range, value) { a, b -> a.mergeWith(b) }
+            }
         }
+
+        putRange(
```

Comment:
These 3 method calls might need a comment to clarify that we're marking the 3 types of classes as visited with their corresponding parameters.

Reply:
Fixed in 0c57be6390f29f51b60b1078138c60211d4dfd1f